Time marker

stormpython opened the pull request:
Closes #2757.

Add option to show current time on vertical bar, line, and area charts.

rashidkpc commented on src/kibana/plugins/vis_types/controls/vislib_basic_options.html:
```diff
@@ -17,4 +17,10 @@
+  <div class="vis-option-item">
+    <label>
+      <input type="checkbox" ng-model="vis.params.addTimeMarker" ng-checked="vis.params.addTimeMarker">
+      Add time marker
```
This could be more descriptive. Maybe "Current time marker"

rashidkpc commented on src/kibana/components/vislib/visualizations/time_marker.js:
```diff
@@ -0,0 +1,41 @@
+      this.xScale = xScale;
+      this.height = height;
+      this.lineClass = 'time-marker';
+      this.time = new Date().getTime();
```
Can you add an option to pass this in, preferably having it take an array so we can pass in multiple?

rashidkpc replied:
Also, it might be a good idea to have this take a date math string

rashidkpc commented:
Can you merge master on this one?

rashidkpc commented:
I noticed if you hover the bar you can brush from that point

Also, can you change the default color to something a bit less jarring? Perhaps rgba(200,0,0,0.3):

![screen shot 2015-04-09 at 4 18 02 pm](https://cloud.githubusercontent.com/assets/1250387/7079019/2356123a-ded4-11e4-9be8-d065f6b4e6e4.png)

rashidkpc commented:
Can you also enable this by default in Discover?

rashidkpc commented on src/kibana/components/vislib/visualizations/time_marker.js:
```diff
@@ -0,0 +1,62 @@
+      this.xScale = xScale;
+      this.height = height;
+      this.lineClass = 'time-marker';
+      this.stroke = 'blue';
```
rgba(200,0,0,0.3) perhaps more pleasing here

stormpython commented:
@rashidkpc addressed your comments.

rashidkpc commented:
I can see a couple things that would make this better, but maybe we should save them for phase 2? Perhaps a new ticket?

- Specifying the color of the marker for individual markers in the *times* array
- Configuring a tooltip for each marker in the times array

I can't decide if it makes sense to have the the times array and addTimeMarker be different things. I think it does. @spalger @w33ble opinions?

Pros to keeping them separate:
- Easy to implement the checkbox in Visualize
- If you want both a current time marker, and an arbitrary number of other markers you don't have to to worry about keeping one of the items in the array up to date

Cons:
- Leads to duplication in places. Perhaps not a big deal.

rashidkpc commented:
I'm going to assign this over to @w33ble for more input.

w33ble commented on src/kibana/components/vislib/visualizations/line_chart.js:
```diff
@@ -291,6 +296,7 @@ define(function (require) {
 
           width = elWidth - margin.left - margin.right;
           height = elHeight - margin.top - margin.bottom;
+          timeMarker = new TimeMarker(times, xScale, height);
```
Why is the timeMarker being newed up here if there's a chance it's not going to be used? Why not new it up in the `if (addTimeMarker)` section?

w33ble commented on src/kibana/components/vislib/visualizations/time_marker.js:
```diff
@@ -0,0 +1,58 @@
+      });
+    };
+
+    TimeMarker.prototype.get = function (field) {
```
Nothing on `this` is hidden, there's no need for a getter here

w33ble replied:
A generic getter isn't useful here - all those properties are available on the instance, and it doesn't prevent needing outside knowledge of internal variable names. If you need a specific getter, make a method for it. It looks like this is mostly used for testing though, so can be removed.

w33ble commented on src/kibana/components/vislib/visualizations/time_marker.js:
```diff
@@ -0,0 +1,58 @@
+      return this[field] ? this[field] : undefined;
+    };
+
+    TimeMarker.prototype.set = function (field, val) {
```
A generic setter seems wrong here. The only thing that's not being set when newing up is `lineClass`, and that's the only thing in the tests as well. Why not add a `setClass` or `setLineClass` method that does this, then you don't need any outside knowledge of the line class variable name.

w33ble commented on src/kibana/components/vis/vis.js:
```diff
@@ -102,6 +102,13 @@ define(function (require) {
+    Vis.prototype.hasSchemaAgg = function (schemaName, aggTypeName) {
+      var aggs = this.aggs.bySchemaName[schemaName] || [];
+      return aggs.some(function (agg) {
+        return agg.type.name === aggTypeName;
```
`agg.type` is not always defined, so this is causing errors.

stormpython commented:
@rashidkpc @w33ble 

I tried to address all of your concerns as well as comments for improvement.

1. The `TimeMarker` class is only instantiated when `addTimeMarker` is **true** for all point series charts.
2. Removed generic getters/setters.
3. Fixed bug with `hasSchemaAgg`. @spalger can you review this particular file as I am not sure what other things use `hasSchemaAgg`. It is being used by the `vislib_basic_options` html file to determine whether to show the check box for adding time markers. if `hasSchemaAgg` evaluates to true it is shown.
4. To address future proofing, I have changed the times array input to be an array of objects. The times array API is as follows:

```javascript
[
  {
    time: 'now-5m', // dateMathString, timestamp, etc
    class: 'time-marker', //  (optional) class for the line that is created, defaults to 'time-marker'
    color: '#c80000', //  (optional) hex color for the line, defaults to '#c80000'
    opacity: 0.3, // (optional) opacity of the line (number), defaults to 0.3
    width: 2 // (optional) width of line, defaults to 2px
  }
];
```

I think this should be reviewed by both @rashidkpc and @w33ble again due to the types of changes I've made.

spalger commented on src/kibana/components/vis/vis.js:
```diff
@@ -102,6 +102,14 @@ define(function (require) {
+    Vis.prototype.hasSchemaAgg = function (schemaName, aggTypeName) {
+      var aggs = this.aggs.bySchemaName[schemaName] || [];
+      return aggs.some(function (agg) {
+        if (!agg.type || !agg.type.name) return false;
```
LGTM. It is quite possible for `agg.type` to be undefined. `type.name` on the other hand should never be undefined, but doesn't hurt to check.
